route-rs-runtime/src/link/tee_link.rs

```diff
@@ -1,50 +1,103 @@
 use crate::link::task_park::*;
-use crate::link::{AsyncEgressor, PacketStream};
+use crate::link::{AsyncEgressor, Link, LinkBuilder, PacketStream};
 use crossbeam::atomic::AtomicCell;
 use crossbeam::crossbeam_channel;
 use crossbeam::crossbeam_channel::{Receiver, Sender};
 use futures::{Async, Future, Poll, Stream};
 use std::sync::Arc;
 
-pub struct TeeLink<P: Sized + Clone> {
-    pub ingressor: TeeIngressor<P>,
-    pub egressors: Vec<AsyncEgressor<P>>,
+#[derive(Default)]
+pub struct TeeLinkBuilder<Packet: Sized + Clone + Send> {
+    in_stream: Option<PacketStream<Packet>>,
+    queue_capacity: usize,
+    branches: Option<usize>,
 }
 
-impl<P: Sized + Clone> TeeLink<P> {
-    pub fn new(input_stream: PacketStream<P>, queue_capacity: usize, branches: usize) -> Self {
+impl<Packet: Sized + Clone + Send> TeeLinkBuilder<Packet> {
+    pub fn new() -> Self {
+        TeeLinkBuilder {
+            in_stream: None,
+            queue_capacity: 10,
+            branches: None,
+        }
+    }
+
+    /// Changes queue_capacity, default value is 10.
+    /// Valid range is 1..=1000
+    pub fn queue_capacity(self, queue_capacity: usize) -> Self {
+        assert!(
+            queue_capacity <= 1000,
+            format!("queue_capacity: {} > 1000", queue_capacity)
+        );
+        assert_ne!(queue_capacity, 0, "queue capacity must be non-zero");
+
+        TeeLinkBuilder {
+            in_stream: self.in_stream,
+            queue_capacity,
+            branches: self.branches,
+        }
+    }
+
+    pub fn branches(self, branches: usize) -> Self {
         assert!(
             branches <= 1000,
             format!("Tee Link branches: {} > 1000", branches)
         );
-        assert!(
-            queue_capacity <= 1000,
-            format!("Tee Link queue_capacity: {} > 1000", queue_capacity)
+        assert_ne!(branches, 0, "branches must be non-zero");
+
+        TeeLinkBuilder {
+            in_stream: self.in_stream,
+            queue_capacity: self.queue_capacity,
+            branches: Some(branches),
+        }
+    }
+}
+
+impl<Packet: Sized + Send + Clone + 'static> LinkBuilder<Packet, Packet>
```

Comment:
Eventually the trait constraints on what can be passed through `Link`s might get a little out of hand, so maybe later we should alias those into another trait:

```
trait LinkPassable: Sized + Send + Clone + 'static
```

so we can write this definition as

```
impl<P: LinkPassable> LinkBuilder<P, P>
```

Reply:
Yeah, something like that.  I  want to talk about this in person. Since one we can't make packets 'static, and two we should reevaluate what our Trait requirements are in general.

Reply:
I did go through and remove the Sized requirement, since Clone requires Sized.